### DISCUSSION TITLE
Component based approach

### DISCUSSION THREAD

**Discussion by ai** [wprig/wprig#756] [Ideas]
## Issue Overview

Let’s replace `dev/css/footer.css` and `dev/js/footer.js` to `dev/components/footer/index.js` and `dev/components/footer/index.css`

## Describe your environment

@mor10 after his great talk asked me to check this amazing project for the best practices in CSS development.

Right now there is a very popular practice to replace `css/` and `js/` dirs to components. Every file of the component will be in the single dir.

Benefits:

1. Much easy to find dead code. In the current approach, it is easy to remove  `dev/js/footer.js` and forget to remove `dev/css/footer.css`. In components approach you remove component dir.
2. If you change something in JS file, you will change CSS of the same component. So it is better to keep them together.
3. It will force better isolation between components.

Right now this approach is very popular. For instance, it is recommended by BEM and Create React App.

Unfortunately, it will be the major change.

**Comment by mor10** [comment]
I'm wondering if we can extend this idea to include PHP files as well. That's sort of the thinking here anyway: contain each component in its own folder so they are easy to isolate and move around.

**Comment by ai** [comment]
Yeap, it will be perfect. PHP templates to generate HTML in the same for with CSS for this HTML.

**Comment by mor10** [comment]
I'd love for you to take a look at the `v2.0` branch now @ai. PHP is component-based, CSS is managed in a new and very different way through PostCSS, and there are other major changes as well. Once 2.0 is out, we're switching focus to rethinking CSS from the ground up, which will be 3.0 most likely.

**Comment by ai** [comment]
Hi. I will try to look on next week.

**Comment by ai** [comment]
> https://github.com/wprig/wprig/blob/v2.0/.browserslistrc

I recommend `defaults`. It also contains `not dead` (to remove dead browsers) and `firefox esr` (important for enterprise users).

> https://github.com/wprig/wprig/blob/v2.0/assets/css/src/comments.css#L49

:+1: for CSS Custom Properties

> https://github.com/wprig/wprig/blob/v2.0/assets/css/src/comments.css#L118

The strange new line. Should we add a newline before `}`? (and fix Stylelint config)

> https://github.com/wprig/wprig/blob/v2.0/assets/css/src/content.css#L57-L59

I found too many `stylelint disable`. Should we change config, send issues to Stylelint or maybe add description to every stylelint disable?

> https://github.com/wprig/wprig/blob/v2.0/assets/css/src/reset.css

Should we use [postcss-normalize](https://github.com/csstools/postcss-normalize). It is like “Autoprefixer” for Normalize.css (adds only necessary resets for your target browsers).

**Comment by lxbdr** [comment]
I recently built a little component-based framework for myself and realized that it's not a good idea to name all the files `index`. Makes it easy to load the necessary files as you expect a certain name but it is not really convenient to work with these files in an IDE because you try to find a file via search but all of them are named `index`
